src/Data/Matrix.purs

```diff
@@ -82,48 +124,77 @@ instance applyMat :: Apply (Mat s) where
-getElem i j m@(Mat l) = unsafePartial $ fromJust (l !! (i * sized (Proxy :: Proxy s) + j))
+getElem :: forall r c a.
+  Sized r => Sized c =>
+  Mat r c a     -- Matrix
+  -> Int        -- Row
+  -> Int        -- Column
+  -> a
+getElem m@(Mat l) i j = unsafePartial $ fromJust (l !! index m i j)
 
 -- | Scale a matrix by a given factor.
 --   Example:
 --
 -- >               ( 1 2 3 )   (  2  4  6 )
 -- >               ( 4 5 6 )   (  8 10 12 )
 -- > scaleMatrix 2 ( 7 8 9 ) = ( 14 16 18 )
-scaleMatrix ::  forall a s. (EuclideanRing a) => a -> Mat s a -> Mat s a
+scaleMatrix ::  forall r c a. (EuclideanRing a) => a -> Mat r c a -> Mat r c a
 scaleMatrix = (<$>) <<< (*)
 
-fromArray :: forall a s. (Sized s) => Array a -> Mat s a
-fromArray l =
-  let size = sized (Proxy :: Proxy s)
-  in case size * size of
+fromArrayColumns :: forall r c a. Sized r => Sized c => Array a -> Mat r c a
+fromArrayColumns l =
+  case cs * rs of
         i | i == length l -> Mat l
-          | otherwise     -> fail "Matrix>>fromArray: Wrong array length!"
+          | otherwise     -> fail "Matrix>>fromArrayColumns: Wrong array length!"
+  where
+    cs = sized (Proxy :: Proxy c)
+    rs = sized (Proxy :: Proxy r)
+
+fromArrayRows :: forall r c a. Sized r => Sized c => Array a -> Mat r c a
+fromArrayRows = transpose <<< fromArrayColumns
+
+toArrayColumns :: forall r c a. Mat r c a -> Array a
+toArrayColumns (Mat a) = a
 
-toArray :: forall s a. Mat s a -> Array a
-toArray (Mat a) = a
+toArrayRows :: forall r c a. Sized r => Sized c => Mat r c a -> Array a
+toArrayRows = toArrayColumns <<< transpose
+
+fromColumns :: forall r c a. Sized r => Sized c => Array (V.Vec r a) -> Mat r c a
+fromColumns vs | ((length vs) == sized (Proxy :: Proxy c)) = Mat $ concat $ V.toArray <$> vs
+fromColumns _ = fail "Matrix>>fromColumns: Wrong array length!"
+
+fromRows :: forall r c a. Sized r => Sized c => Array (V.Vec c a) -> Mat r c a
+fromRows vs | ((length vs) == sized (Proxy :: Proxy r)) = transpose $ Mat $ concat $ V.toArray <$> vs
+fromRows _ = fail "Matrix>>fromColumns: Wrong array length!"
 
 -- | /O(rows*cols)/. The transpose of a matrix.
 --   Example:
 --
 -- >           ( 1 2 3 )   ( 1 4 7 )
 -- >           ( 4 5 6 )   ( 2 5 8 )
 -- > transpose ( 7 8 9 ) = ( 3 6 9 )
-transpose :: forall a s.  (Sized s) => Mat s a -> Mat s a
-transpose m = generate $ \ i j -> getElem j i m
+transpose :: forall r c a.  Sized r => Sized c => Mat r c a -> Mat c r a
+transpose m = generate $ \i j -> getElem m j i
+
+mulmatvec :: forall r c. Sized r => Sized c => Mat r c Number -> V.Vec c Number -> V.Vec r Number
+mulmatvec m v = V.Vec $ V.dot <$> rows m <*> [v]
+
+mulvecmat :: forall r c. Sized r => Sized c => V.Vec r Number -> Mat r c Number -> V.Vec c Number
+mulvecmat v m = V.Vec $ V.dot <$> [v] <*> columns m
+
+mulmatmat :: forall rl u cr. Sized rl => Sized u => Sized cr => Mat rl u Number -> Mat u cr Number -> Mat rl cr Number
+mulmatmat l r = fromColumns $ (mulmatvec l <$> columns r)
```

Comment:
These could also be generalized with `Semiring`, with upstream changes I guess ...